### PR TITLE
Closes #19147: Move startup metrics to right after Glean initialization

### LINKED ISSUE #19147: Investigate into null values in startup metrics
Currently there are large number of startup metrics are reported as null.  Investigate into how startup metrics are reported to understand the cause/solution of the issue.

Example query: https://sql.telemetry.mozilla.org/queries/79505/

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by gabrielluong** [comment]
Is there also a data eng/sci peer that can take a look at this since we are also updating the glean types for some metrics?

**Comment by rocketsroger** [comment]
> Is there also a data eng/sci peer that can take a look at this since we are also updating the glean types for some metrics?

Sure, I'll add reviewer from data science.

**Comment by rocketsroger** [comment]
@Dexterp37 please confirm that this change is what we still want for https://github.com/mozilla-mobile/fenix/issues/12131.  Thanks,

File: app/metrics.yaml
Comment (badboy) [inline app/metrics.yaml:1273]: Just for your awareness: This will work, but land in a different column.
Old: `metrics.string_list.preferences_show_search_suggestions`
New: `metrics.boolean.preferences_show_search_suggestions`

That means comparing these across versions is a bit more involved, but not impossible.
(There should never be a client that has both set though)

(I'm still waiting for confirmation that what I said is actually true)

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (badboy) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:521]: Oh I see now that is just copied from the old code.
Still wonder what this part means.

**Review by badboy — CHANGES_REQUESTED** [review]
Blocking here while I wait for confirmation that just changing the type won't break anything.

This is now code that's executed earlier, is there any way to verify that it has a neglibile impact on startup?

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:521]: Yes, this was copied from the original function.  I will update it so it makes more sense.  Thanks,

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:522]: Sure, I'll update this since I'll rewrite this.  Not sure why originally this was double spaced.

File: app/metrics.yaml
Comment (rocketsroger) [reply app/metrics.yaml:1273]: Yes, please confirm that this is OK.  Since this workaround is because currently there are high number of startup metrics that have null values, I would argue that comparing against data we collected before is not very useful.

**Comment by rocketsroger** [comment]
> This is now code that's executed earlier, is there any way to verify that it has a neglibile impact on startup?

Yes, I think this will have some impact on startup performance.  However, this is currently the only workaround suggestion from the Glean team.  Once the Glean investigation is over we will update this if there is a better solution.

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:152]: Agreed.  I will move this into the `setStartupMetrics`.  Thanks,

**Comment by gabrielluong** [comment]
Will take another look at this thoroughly tomorrow.

File: app/metrics.yaml
Comment (mdboom) [reply app/metrics.yaml:1273]: Yeah, I agree it sounds like it's worthwhile to make the change.  You still may want to add in the description field "Prior to release X, this metric was of type `string_list`" so that *if* someone wanted to trace back through the (problematic) historical data they'd at least know how...

**Comment by gabrielluong** [comment]
Sorry, didn't get to this today, will review it tomorrow!

File: app/metrics.yaml
Comment (acmiyaguchi) [reply app/metrics.yaml:1273]: This kind of change is not supported by our tooling, a similar change of a string to a datetime in the past has caused delays on our schema deploys: https://github.com/mozilla/mozilla-schema-generator/issues/118

Type changes is typically frowned upon at a low level because it [breaks schema evolution rules](https://cloud.google.com/bigquery/docs/managing-table-schemas) that allows for low operational overhead during ingestion time. It's possible to support changing types in the schema generator based on the behavior of moving the probe to a new namespace (which avoids a schema incompatible change). However, this will be confusing from a metadata perspective since we lose context on the type change.

I suggest expiring these probes and creating a new probe with the correct type, while clearly documenting the reason for deprecation so that tools like the [glean dictionary](https://dictionary.telemetry.mozilla.org/) can pick this up.

Comment (gabrielluong) [reply app/metrics.yaml:1273]: @rocketsroger Can you ping me when you think you need a final review? Thanks!

Comment (rocketsroger) [reply app/metrics.yaml:1273]: @gabrielluong Sure, this should be the final change but I'll ping you once I have the go ahead from the data science team.  Thanks,

**Review by badboy — APPROVED** [review]
Looks good to me now!

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/FenixApplication.kt:609]: I imagine we should change this comment if not remove it entirely since they're boolean now. I think this comment shouldn't be applied to the `with` block either since it only applies to line 613-623.

**Review by gabrielluong — APPROVED** [review]
Do we need to check with the perf-team?

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:609]: Good catch.   Thanks I'll update

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt:848]: Sure, I'll try to describe this service.

**Comment by rocketsroger** [comment]
> Do we need to check with the perf-team?

I'll notify the perf-team.  This will affect the startup performance but unfortunately there's no other solutions.

**Pull request merged**
